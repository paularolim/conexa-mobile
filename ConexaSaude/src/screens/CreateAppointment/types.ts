import { Control } from 'react-hook-form';
import * as yup from 'yup';

import { schema } from './validation';

export type CreateAppointmentSchema = yup.InferType<typeof schema>;

export interface CreateAppointmentViewProps {
  loading: boolean;
  errorMessage?: string;
  control: Control<CreateAppointmentSchema, any>;
  onSubmit: (e?: React.BaseSyntheticEvent<object, any, any> | undefined) => Promise<void>;
  handleDismissKeyboard: () => void;
}
