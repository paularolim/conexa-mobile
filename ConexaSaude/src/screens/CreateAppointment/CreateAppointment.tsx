import { View } from 'react-native';

import { Text } from '@components/Text';

export function CreateAppointment() {
  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text>CreateAppointment</Text>
    </View>
  );
}
