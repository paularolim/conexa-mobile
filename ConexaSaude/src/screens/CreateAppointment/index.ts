export * from './CreateAppointment';
export * from './CreateAppointmentView';
export * from './types';
export * from './useCreateAppointmentViewModel';
export * from './validation';
