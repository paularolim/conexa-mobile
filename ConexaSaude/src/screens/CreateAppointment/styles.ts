import { View } from 'react-native';
import styled from 'styled-components/native';

import { Text } from '@components/Text';

export const Container = styled(View)`
  background-color: ${({ theme }) => theme.colors.white};
  flex: 1;
  padding: 12px 24px;
  gap: 12px;
`;

export const Form = styled(View)`
  flex: 1;
  gap: 8px;
`;

export const Title = styled(Text)`
  font-size: 20px;
  font-weight: bold;
  color: ${({ theme }) => theme.colors.gray};
`;
