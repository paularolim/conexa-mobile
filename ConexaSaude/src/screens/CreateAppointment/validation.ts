import * as yup from 'yup';

export const schema = yup
  .object({
    patient: yup.string().required('Nome do paciente não informado!'),
    date: yup.string().required('Data não informada!'),
    observation: yup.string().required('Observação não informada!'),
  })
  .required();
