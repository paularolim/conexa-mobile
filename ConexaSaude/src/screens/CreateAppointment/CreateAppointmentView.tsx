import { Controller } from 'react-hook-form';
import { ActivityIndicator, TouchableWithoutFeedback } from 'react-native';

import { Alert } from '@components/Alert';
import { Button } from '@components/Button';
import { Input } from '@components/Input';

import { Container, Form, Title } from './styles';
import { CreateAppointmentViewProps } from './types';

export function CreateAppointmentView({
  control,
  loading,
  errorMessage,
  onSubmit,
  handleDismissKeyboard,
}: CreateAppointmentViewProps) {
  return (
    <TouchableWithoutFeedback onPress={handleDismissKeyboard}>
      <Container>
        <Title>Criar consulta</Title>
        <Form>
          {errorMessage && <Alert type="error" message={errorMessage} />}
          <Controller
            name="patient"
            control={control}
            render={({ field: { value, onChange } }) => (
              <Input label="Nome do paciente" value={value} onChangeText={onChange} />
            )}
          />
          {/* TODO: improve ux for date input */}
          <Controller
            name="date"
            control={control}
            render={({ field: { value, onChange } }) => (
              <Input label="Data da consulta" value={value} onChangeText={onChange} />
            )}
          />
          <Controller
            name="observation"
            control={control}
            render={({ field: { value, onChange } }) => (
              <Input label="Observação" value={value} onChangeText={onChange} />
            )}
          />
        </Form>
        {loading ? (
          <ActivityIndicator testID="activity-indicator" />
        ) : (
          <Button onPress={onSubmit}>Salvar</Button>
        )}
      </Container>
    </TouchableWithoutFeedback>
  );
}
