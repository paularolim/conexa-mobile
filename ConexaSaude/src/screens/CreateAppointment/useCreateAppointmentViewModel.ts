import { yupResolver } from '@hookform/resolvers/yup';
import { useMemo } from 'react';
import { useForm } from 'react-hook-form';
import { Alert } from 'react-native';

import { useCreateAppointment } from '@hooks/useCreateAppointment';

import { CreateAppointmentSchema } from './types';
import { schema } from './validation';

export function useCreateAppointmentViewModel() {
  const { fetchCreateAppointment, error, loading } = useCreateAppointment();

  const {
    handleSubmit,
    formState: { errors },
    control,
    reset,
  } = useForm<CreateAppointmentSchema>({ resolver: yupResolver(schema) });

  const errorMessage = useMemo(
    () =>
      errors.patient?.message ||
      errors.date?.message ||
      errors.observation?.message ||
      (error ? 'Não foi possível salvar a consulta.' : null),
    [error, errors.date?.message, errors.observation?.message, errors.patient?.message],
  );

  const onSubmit = handleSubmit((data) => {
    fetchCreateAppointment(data).then(() => {
      reset();
      Alert.alert('Sucesso', 'A consulta foi salva!');
    });
  });

  return { errorMessage, control, loading, onSubmit };
}
