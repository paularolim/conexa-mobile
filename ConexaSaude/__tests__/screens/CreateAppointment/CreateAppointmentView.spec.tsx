import { render, renderHook } from '@testing-library/react-native';
import { useForm } from 'react-hook-form';
import renderer from 'react-test-renderer';

import { CreateAppointmentView } from '@screens/CreateAppointment';

import { Wrapper } from '../../Wrapper';

jest.useFakeTimers().setSystemTime(new Date('2023-10-04T12:21:45.926Z'));

describe('CreateAppointmentView', () => {
  it('should render screen correctly', () => {
    const { result } = renderHook(() =>
      useForm<{ patient: string; date: Date; observation: string }>(),
    );

    const tree = renderer
      .create(
        <Wrapper>
          <CreateAppointmentView
            loading={false}
            control={result.current.control}
            onSubmit={jest.fn()}
            handleDismissKeyboard={jest.fn()}
            errorMessage={null}
            isDateModalOpen={false}
            handleToggleDateModal={jest.fn()}
          />
        </Wrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render screen correctly on loading', () => {
    const { result } = renderHook(() =>
      useForm<{ patient: string; date: Date; observation: string }>(),
    );

    const { getByTestId } = render(
      <Wrapper>
        <CreateAppointmentView
          loading
          control={result.current.control}
          onSubmit={jest.fn()}
          handleDismissKeyboard={jest.fn()}
          errorMessage={null}
          isDateModalOpen={false}
          handleToggleDateModal={jest.fn()}
        />
      </Wrapper>,
    );
    expect(getByTestId('activity-indicator')).toBeTruthy();
  });

  it('should render screen correctly on loading', () => {
    const { result } = renderHook(() =>
      useForm<{ patient: string; date: Date; observation: string }>(),
    );

    const { getByText } = render(
      <Wrapper>
        <CreateAppointmentView
          errorMessage="Error message"
          loading={false}
          control={result.current.control}
          onSubmit={jest.fn()}
          handleDismissKeyboard={jest.fn()}
          isDateModalOpen={false}
          handleToggleDateModal={jest.fn()}
        />
      </Wrapper>,
    );
    expect(getByText('Error message')).toBeTruthy();
  });
});
