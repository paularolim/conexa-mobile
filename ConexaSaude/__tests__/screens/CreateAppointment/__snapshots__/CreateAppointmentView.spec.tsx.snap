// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAppointmentView should render screen correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "gap": 12,
        "paddingBottom": 12,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 12,
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
        },
        {
          "color": "#707683",
          "fontSize": 20,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Criar consulta
  </Text>
  <View
    style={
      [
        {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "gap": 8,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#FFFFFF",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000000",
            },
            {
              "color": "#707683",
              "fontSize": 18,
            },
          ]
        }
      >
        Nome do paciente
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#707683",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flexDirection": "row",
              "minHeight": 48,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
              "width": "100%",
            },
          ]
        }
      >
        <TextInput
          onChangeText={[Function]}
          style={
            [
              {
                "color": "#000000",
                "flexGrow": 1,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
            ]
          }
          testID="field"
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#FFFFFF",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000000",
            },
            {
              "color": "#707683",
              "fontSize": 18,
            },
          ]
        }
      >
        Data da consulta
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#707683",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flexDirection": "row",
              "minHeight": 48,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
              "width": "100%",
            },
          ]
        }
      >
        <TextInput
          onChangeText={[Function]}
          style={
            [
              {
                "color": "#000000",
                "flexGrow": 1,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
            ]
          }
          testID="field"
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-start",
            "backgroundColor": "#FFFFFF",
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000000",
            },
            {
              "color": "#707683",
              "fontSize": 18,
            },
          ]
        }
      >
        Observação
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#707683",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flexDirection": "row",
              "minHeight": 48,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
              "width": "100%",
            },
          ]
        }
      >
        <TextInput
          onChangeText={[Function]}
          style={
            [
              {
                "color": "#000000",
                "flexGrow": 1,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
            ]
          }
          testID="field"
        />
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2ED47A",
        "borderRadius": 4,
        "justifyContent": "center",
        "maxWidth": "100%",
        "minHeight": 48,
        "opacity": 1,
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#000000",
          },
          {
            "color": "#FFFFFF",
            "fontSize": 18,
          },
        ]
      }
      type="filled"
    >
      Salvar
    </Text>
  </View>
</View>
`;
